test/controllers/api_users/permissions_controller_test.rb

```diff
@@ -63,64 +126,95 @@ class ApiUsers::PermissionsControllerTest < ActionController::TestCase
+      application = create(:application)
+      api_user = create(:api_user)
 
-      application = create(:application, retired: true)
+      patch :update, params: { api_user_id: api_user, application_id: application }
+
+      assert_not_authenticated
+    end
+
+    should "prevent unauthorised users" do
+      application = create(:application)
       api_user = create(:api_user)
+      create(:access_token, application:, resource_owner_id: api_user.id)
 
-      assert_raises(ActiveRecord::RecordNotFound) do
-        get :edit, params: { api_user_id: api_user, application_id: application }
-      end
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      stub_policy current_user, api_user, update?: false
+
+      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [] } }
+
+      assert_not_authorised
     end
 
-    should "exclude applications with revoked access tokens" do
-      sign_in create(:superadmin_user)
+    should "raise an exception if the user doesn't have access to the application" do
+      sign_in create(:admin_user)
 
       application = create(:application)
       api_user = create(:api_user)
-      create(:access_token, resource_owner_id: api_user.id, application:, revoked_at: Time.current)
 
       assert_raises(ActiveRecord::RecordNotFound) do
-        get :edit, params: { api_user_id: api_user, application_id: application }
+        patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [] } }
       end
     end
 
-    should "include applications with revoked access tokens when there is at least one non-revoked access token" do
+    should "raise an exception if the application is retired" do
       sign_in create(:superadmin_user)
 
-      application = create(:application)
-      api_user = create(:api_user)
-      create(:access_token, resource_owner_id: api_user.id, application:, revoked_at: Time.current)
-      create(:access_token, resource_owner_id: api_user.id, application:)
-
-      get :edit, params: { api_user_id: api_user, application_id: application }
+      application = create(:application, retired: true)
+      permission = create(:supported_permission, application:)
+      api_user = create(:api_user,
+                        with_signin_permissions_for: [application],
+                        with_permissions: { application => [permission.name] })
+      create(:access_token, application:, resource_owner_id: api_user.id)
 
-      assert_response :success
+      assert_raises(ActiveRecord::RecordNotFound) do
+        patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [application.id] } }
+      end
     end
 
-    should "raise an exception if the user does not have an access token for the application" do
+    should "raise an exception if the user cannot be found" do
```

Comment:
I think this test can go? It's not called out in your list in the commit message, we don't otherwise test that non-existent params raise a `RecordNotFound` error (it's sort of testing the default Rails behaviour?), and it's not tested for the other actions. I think I removed this in the equivalent commit on the other PR